### ISSUE TITLE
<HOME> or <END> delete content of command input area 

### ISSUE THREAD

**Issue by fbi-ranger** [hercules-390/hyperion#70]
Hercules: Hyperion
Host OS: NONE 

According to http://www.hercules-390.org/hercinst.html#keyboard Keys <HOME> should move the cursor to the begin of the command input area. <END> should move the cursor to the end of the command input area. Having a typed in command in the command input area and pressing either of the keys, the content of the input area is cleared. In contradiction to the description, <END> switches to the semi-graphical "New Panel" display mode. 

This behaviour is quite annoying, especially when the command entered is long.

**Comment by ivan-w** [comment]
Florian,

How the <HOME> and <END> key is processed is entirely dependent on the 
3270 emulator your are using, and is usually processed locally, without 
intervention of the host to which the 3270 session is connected.

--Ivan

**Comment by fbi-ranger** [comment]
Seems to be a misunderstanding, I mean the "HMC" panel of Hercules.

**Comment by Fish-Git** [comment]
I noticed the same thing myself when testing Herc on my Linux system (CentOS 6.4 VMware virtual machine). Our Linux version of console (panel) keyboard handling leave a lot to be desired. It is missing a lot of the functionality that exists for Windows.

**Comment by Fish-Git** [comment]
Further info: I believe we (Hercules developers and possibly even the Hercules community as well, I can't remember) may have discussed this issue before in years past, and as I recall the ultimate answer was we need to basically redesign/rewrite our panel/keyboard handling code (at least for *nix anyway) and replace it with "ncurses" I believe it's called:

* [https://en.wikipedia.org/wiki/Ncurses] (https://en.wikipedia.org/wiki/Ncurses)

Since Linux (non-Windows) programming is not my forte I believe it would be best if someone else handled this issue.

**Comment by ivan-w** [comment]
On 9/27/2015 5:15 AM, Fish-Git wrote:
>
> Further info: I believe we (Hercules developers and possibly even the 
> Hercules community as well, I can't remember) may have discussed this 
> issue before in years past, and as I recall the ultimate answer was we 
> need to basically redesign/rewrite our panel/keyboard handling code 
> (at least for *nix anyway) and replace it with "ncurses" I believe 
> it's called:
>
>   * https://en.wikipedia.org/wiki/Ncurses
>
> Since Linux (non-Windows) programming is not my forte I believe it 
> would be best if someone else handled this issue.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/hercules-390/hyperion/issues/70#issuecomment-143514885>.
>
I think "curses" is the norm ("ncurses" is the GNU implementation of 
"curses")...

(But I could be mistaken)

However, my feeling about this is that there should be a clear 
separation between the console/control/management function and the 
"engine" (cpu/channel)

Just throwing ideas up in the air here... (I've been thinking about this 
since... Oh wait.. 1992 when I first thought of a S/370 emulator myself)

--Ivan

**Comment by jphartmann** [comment]
Public Domain Curses is another unix curses implementation.  It can support X11; this has the advantage (at least from my perspective) that it opens a separate X window for the display so that you retain the standard files for debugging.

The flip side is that it attaches the X stuff with fork(), so it needs to be done before creating threads.  The programming interface is mostly the same, but there are differences in character encoding.  (There are betwen all curses implementations.)

**Comment by jphartmann** [comment]
I agree with Ivan about separation.  At best the address spaces of the channel and device manager should be disjunct so that programming errors in a device manager does not bring down a CPU.  (Try throwing 4000 nulls into the TN3270 server--be sure to back up your disks first.)    

On my P390/Linux this was achieved by having subchannel 255 be able to send orders to the channel that affected the card, whereas other subchannels could perform only the OEMI bus/tag operations.  The subchannels were in a virtual file system created by the channel, which was a kernel module also supporting the PCI interface to the card.

For a while I hoped that FUSE would be the the answer, but its lack of a proper handling of IOCTL has made me give up that approach.

**Comment by truda** [comment]
Hello,
a simple workaround to this problem is to start hercules in your TTY eg press ctrl-alt-F2 and logon and start Hercules. Here the keys HOME and END should behave as expected for major linux distributions.
The simplest solution is probably to configure your terminal emulator (XTerm or Gnome Terminal or other) differently. I'll have a look into this soon.
